lib/flat-compat.js

```diff
@@ -0,0 +1,308 @@
+                }
+
+                languageOptions[key] = eslintrcConfig[key].definition;
+                continue;
+            }
+
+            // clone any object values that are in the eslintrc config
+            if (eslintrcConfig[key] && typeof eslintrcConfig[key] === "object") {
+                languageOptions[key] = {
+                    ...eslintrcConfig[key]
+                };
+            } else {
+                languageOptions[key] = eslintrcConfig[key];
+            }
+        }
+    }
+
+    // copy over linterOptions
+    for (const key of linterOptionsKeysToCopy) {
+        if (key in eslintrcConfig && typeof eslintrcConfig[key] !== "undefined") {
+            flatConfig.linterOptions = linterOptions;
+            linterOptions[key] = eslintrcConfig[key];
+        }
+    }
+
+    // move ecmaVersion a level up
+    if (languageOptions.parserOptions) {
+
+        if ("ecmaVersion" in languageOptions.parserOptions) {
+            languageOptions.ecmaVersion = languageOptions.parserOptions.ecmaVersion;
+            delete languageOptions.parserOptions.ecmaVersion;
+        }
+
+        if ("sourceType" in languageOptions.parserOptions) {
+            languageOptions.sourceType = languageOptions.parserOptions.sourceType;
+            delete languageOptions.parserOptions.sourceType;
+        }
+
+        // check to see if we even need parserOptions anymore and remove it if not
+        if (Object.keys(languageOptions.parserOptions).length === 0) {
+            delete languageOptions.parserOptions;
+        }
+    }
+
+    // overrides
+    if (eslintrcConfig.criteria) {
+        flatConfig.files = [absoluteFilePath => eslintrcConfig.criteria.test(absoluteFilePath)];
+    }
+
+    // translate plugins
+    if (eslintrcConfig.plugins && typeof eslintrcConfig.plugins === "object") {
+        debug(`Translating plugins: ${eslintrcConfig.plugins}`);
+
+        flatConfig.plugins = {};
+
+        for (const pluginName of Object.keys(eslintrcConfig.plugins)) {
+
+            debug(`Translating plugin: ${pluginName}`);
+            debug(`Resolving plugin '${pluginName} relative to ${resolvePluginsRelativeTo}`);
+
+            const { definition: plugin, error } = eslintrcConfig.plugins[pluginName];
+
+            if (error) {
+                throw error;
+            }
+
+            flatConfig.plugins[pluginName] = plugin;
+
+            // create a config for any processors
+            if (plugin.processors) {
+                for (const processorName of Object.keys(plugin.processors)) {
+                    if (processorName.startsWith(".")) {
+                        debug(`Assigning processor: ${pluginName}/${processorName}`);
+
+                        configs.unshift({
+                            files: [`**/*${processorName}`],
+                            processor: plugin.processors[processorName]
+                        });
+                    }
+
+                }
+            }
```

Comment:
> At the moment, that part of the code doesn't have access to the config array, 

I think the same way as for the https://github.com/eslint/eslintrc/pull/20/files#r550700979 thread resolves it. That uses `eslintrcArray.pluginEnvironments` to solve envs. This uses `eslintrcArray.pluginProcessors` to solve processors.

> and I'd rather not introduce a new parameter for the sake of a small efficiency gain. What I have here is working correctly, so I'd like to stick with this as a first draft.

If end-users override plugins, it will go wrong.

Currently, the file-extension processors are resolved in the compat layer, but user-defined processors (`e.g., `processor: "my-plugin/a-processor"`) aren't resolved. Therefore, if an end-user overrides the` my-plugin` in their config, the file-extension processors and user-defined processors will come from different plugin instances.

Reply:
I was able to pass in `pluginProcessors` to do this lookup.

Reply:
Thanks, I updated it.

Reply:
Hmm, maybe my concern was not clear.

My concern is consistency rather than performance.
People can define processor settings in their own config files.

```json
{
    "plugins": ["my-plugin"],
    "processor": "my-plugin/a-processor"`
}
```

My concern in this thread is that the compact layer doesn't resolve this setting. Therefore, if an end-user overrides the my-plugin in their flat config, file-extension processors and user-defined processors will come from different plugin instances.

I'd like to make more consistent this behavior. All processors are resolved in the compact layer or not.

Reply:
It’s not possible to override what a plugin refers to. Once “my-plugin” is set to an object, an error will be thrown if you attempt to set “my-plugin” to something else. “my-plugin/a-processor” always will mean the same thing regardless of where it is referenced in a flat config.

If that doesn’t address your concern, please open an issue so we can continue discussing. I don’t want to hold up this PR for an edge case.

Reply:
Oh, sorry. I have misunderstood it.

> Once “my-plugin” is set to an object, an error will be thrown if you attempt to set “my-plugin” to something else.

It should solve my concern.